enhancements/microshift/lvms-topolvm-downstream-replacement.md

```diff
@@ -0,0 +1,269 @@
+an empty deviceClass in the LVMCluster and then coordinating that with the vg-manager.
+We have a prototype implementation that can be pushed into GA in time for the 4.17 release based on
+[OCPEDGE-830](https://issues.redhat.com//browse/OCPEDGE-830) that proves this is possible.
+
+#### Switching from TopoLVM to LVMS
+
+The switch from TopoLVM to LVMS will be done in a single release. We will support
+upgrading from 4.16 to 4.16, and from 4.14 to 4.16 (EUS).
+
+The switch will be done by removing the TopoLVM deployment and installing the LVMS
+deployment. The LVMS deployment will be configured to use the existing deviceClasses
+and storageClasses. There will be leftover images from the TopoLVM deployment that
+can be removed manually from the registry after the upgrade.
+
+There should not be other regressions in the upgrade process, as the LVMS migration
+is transparent to users.
+
+### Risks and Mitigations
+
+There is some risk that the LVMS deployment will not work with the existing
+deviceClasses and storageClasses. We will mitigate this risk by testing the
+upgrade process in CI and by providing a rollback mechanism in case of failure.
+Specific test scenarios that are not handled in the generic LVMS use cases should especially involve
+[Software RAIDs introduced with LVM Raid](https://github.com/topolvm/topolvm/blob/2597dc75821009230af3be4067d377eef4bbe0a4/docs/lvmd.md?plain=1#L28-L36) with custom create options
+This is because in LVMS, these are not supported by default and require a custom deviceClass
+and are now enabled for the first time within LVMS through the MicroShift deployment.
+
+Additionally, we will provide documentation on how to correctly set up an empty
+LVMCluster and how to run debugging procedures in case of failure.
+Any issues discovered during this process will be remediated with custom adjustments.
+
+### Drawbacks
+
+The main draw back of this change is that it might require manual remediation
+if a failure occurs during the upgrade process. Since this is a one-time change,
+we believe that the benefits of moving to LVMS outweigh this potential issue.
+
+Additionally, the change will require some additional testing to ensure that
+the upgrade process works as expected. From now on we will need to issue rebase
+updates for the LVMS deployment in MicroShift and that will require adjustments
+
+## Test Plan
+
+We will add an automated test to CI to upgrade the MicroShift version from 4.15 to 4.16 anyhow.
+In this test, we will verify that the upgrade process works as expected and that the LVMS
+deployment is correctly configured to use the existing deviceClasses and storageClasses.
+
+The MicroShift Storage Tests will run regressions and additional tests to ensure that the
+upgrade process works as expected.
+
+Specifically for LVMS, we will introduce a new set of tests that will verify that the
+LVMCluster is correctly configured and that the vg-manager is running as expected.
```

Comment:
What's missing in the proposal is a reference to greenboot. We need to update the greenboot scripts to make sure new storage is up and running, so tests only take care of functionality checks.

See https://github.com/openshift/microshift/blob/main/packaging/greenboot/microshift-running-check.sh#L33

Reply:
I do not think so. Existing volume group management of MicroShift should still work, meaning that the precreated volume group should work without a problem.

Reply:
Last question: are we going to end up with 2 pods in the `openshift-storage` namespace in the "new" LVMS?

See https://github.com/openshift/microshift/blob/main/packaging/greenboot/microshift-running-check.sh#L142.

In general, if there's any new code necessary for making sure LVMS is correctly configured and running, it should go to the greenboot script, not tests. This is because we want this check to also exist in user environments.

Reply:
yes thats correct. We will run with 1 pod named "lvms-operator" and 1 pod named "vg-manager"

Reply:
I am under the impression that any necessary greenboot adjustments will be catched while doing the PR so any small fixes like this should be caught in the e2e tests I think right?

Reply:
Resolving as any mismatches of greenboot or tests would be covered while implementation is going on